core/src/net/cidr/v4.rs

```diff
@@ -53,20 +94,72 @@ impl Cidr for Ipv4Cidr {
         self.address
     }
 
+    #[inline]
+    fn contains(&self, address: Self::Addr) -> bool {
+        self.prefix == (self.mask & u32::from(address))
+    }
+
+    #[inline]
+    fn contains_ip(&self, ip: IpAddr) -> bool {
+        match ip {
+            IpAddr::V4(ip) => self.contains(ip),
+            _ => false,
+        }
+    }
+
+    #[inline]
+    fn first(&self) -> Self::Addr {
+        Ipv4Addr::from(self.mask & u32::from(self.address))
+    }
+
+    #[inline]
+    fn last(&self) -> Self::Addr {
+        Ipv4Addr::from((self.mask & u32::from(self.address)) | !self.mask)
+    }
+
     #[inline]
     fn length(&self) -> usize {
         self.length
     }
 
     #[inline]
-    fn new(address: Self::Addr, length: usize) -> Result<Self, CidrParseError> {
+    fn hostmask(&self) -> Self::Addr {
+        if let Some(mask) = u32::max_value().checked_shr(self.length as u32) {
+            Ipv4Addr::from(mask)
+        } else {
+            Ipv4Addr::from(0)
+        }
+    }
+
+    #[inline]
+    fn netmask(&self) -> Self::Addr {
+        if let Some(mask) = u32::max_value().checked_shl((IPV4ADDR_BITS - self.length) as u32) {
```

Comment:
we already store the masking bits, shouldn't this just be `Ipv4Addr::from(self.mask)`?